pkg/controllers/namespace/namespace_controller.go

```diff
@@ -0,0 +1,228 @@
+	if err != nil {
+		message := "failed to create VPC CR."
+		r.namespaceError(ctx, obj, message, err)
+		// If create VPC CR failed, put ns create event back to queue.
+		return nil, err
+	}
+
+	changes := map[string]string{
+		NS_ANNO_VPC_ERROR: "",
+	}
+	r.updateNSAnnotation(ctx, obj, changes)
+	log.Info("create vpc cr", "VPC", vpc_cr.Name, "for namespace", vpc_cr.Namespace)
+	return vpc_cr, nil
+}
+
+func (r *NamespaceReconciler) namespaceError(ctx *context.Context, k8s_obj client.Object, msg string, err error) {
+	log_err := util.If(err == nil, errors.New(msg), err).(error)
+	log.Error(log_err, msg)
+	changes := map[string]string{NS_ANNO_VPC_ERROR: msg}
+	r.updateNSAnnotation(ctx, k8s_obj, changes)
+}
+
+/*
+	VPC creation strategy:
+
+We suppose namespace should have following annotations:
+  - "nsx.vmware.com/vpc_name": "kube-system/<Supervisor ID>"
+    If the ns contains this annotation, first check if	the namespace in annotation is the same as
+    the one in ns event, if yes, create an infra VPC for it. if	not, skip the whole ns event as the infra
+    VPC will be created its corresponding ns creation event.
+  - "nsx.vmware.com/vpc_network_config":"<Supervisor ID>"
+    If ns do not contains "nsx.vmware.com/vpc_name" annotation. Use this annotation to handle VPC creation.
+    VPC will locate the network config with the CR name, and create VPC using its config.
+  - If the ns do not have either of the annotation above, then we believe it is using default VPC, try to search
+    default VPC in network config CR store. The default VPC network config CR's name is "default".
+*/
+func (r *NamespaceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	obj := &v1.Namespace{}
+	log.Info("reconciling K8s namespace resources", "namespace", req.NamespacedName)
+	metrics.CounterInc(r.NSXConfig, metrics.ControllerSyncTotal, common.MetricResTypeNamespace)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch namespace.", "req", req.NamespacedName)
+		return common.ResultRequeueAfter5mins, err
+	}
+
+	// processing create/update event
+	ns := obj.GetName()
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.NSXConfig, metrics.ControllerUpdateTotal, common.MetricResTypeNamespace)
+		log.Info("start processing namespace create event", "namespace", ns)
+		ctx := context.Background()
+		annotations := obj.GetAnnotations()
+		// read anno "nsx.vmware.com/vpc_name", if ns contains this annotation, it means it will share
+		// infra VPC, if the ns in the annotation is the same as ns event, create infra VPC, if not,
+		// skip the event.
+		nc_name, nc_exist := annotations[NS_ANNO_VPC_NC_NAME]
+		vpc_name, name_exist := annotations[NS_ANNO_VPC_NAME]
+		if name_exist {
+			log.Info("read ns annotation vpc_name", "VPC_NAME", vpc_name)
+			res := strings.Split(vpc_name, "/")
+			// The format should be namespace/vpc_name
```

Comment:
Since it is namespace's anno, isn't it redundant to add namespace/vpc_name instead of vpc_name directly?Then there is need to check if ns == res[0] .